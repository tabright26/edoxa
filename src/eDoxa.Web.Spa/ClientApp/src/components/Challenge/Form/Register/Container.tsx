import { connect } from "react-redux";
import {
  loadChallenge,
  registerChallengeParticipant
} from "store/actions/challenge";
import {
  ChallengesActions,
  REGISTER_CHALLENGE_PARTICIPANT_SUCCESS,
  REGISTER_CHALLENGE_PARTICIPANT_FAIL
} from "store/actions/challenge/types";
import Register from "./Register";
import { ChallengeId } from "types";
import { RouteComponentProps, withRouter } from "react-router-dom";
import { toastr } from "react-redux-toastr";
import { compose } from "recompose";

interface Params {
  readonly challengeId?: ChallengeId;
}

interface OwnProps extends RouteComponentProps<Params> {}

const mapDispatchToProps = (dispatch: any, ownProps: OwnProps) => {
  return {
    registerChallengeParticipant: () =>
      dispatch(
        registerChallengeParticipant(ownProps.match.params.challengeId)
      ).then((action: ChallengesActions) => {
        switch (action.type) {
          case REGISTER_CHALLENGE_PARTICIPANT_SUCCESS: {
            return loadChallenge(ownProps.match.params.challengeId);
          }
          case REGISTER_CHALLENGE_PARTICIPANT_FAIL: {
            return toastr.error(
              "ERROR",
              "An error occurred while registering as a challenge participant."
            );
          }
          default: {
            return Promise.resolve();
          }
        }
      })
  };
};

const enhance = compose<any, any>(
  withRouter,
  connect(null, mapDispatchToProps)
);

export default enhance(Register);
