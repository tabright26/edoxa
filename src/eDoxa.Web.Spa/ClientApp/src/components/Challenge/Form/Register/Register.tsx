import React, { FunctionComponent } from "react";
import { Form } from "reactstrap";
import { reduxForm, InjectedFormProps } from "redux-form";
import Button from "components/Shared/Button";
import { REGISTER_CHALLENGE_PARTICIPANT_FROM } from "utils/form/constants";
import { compose } from "recompose";
import { registerChallengeParticipant } from "store/actions/challenge";
import {
  ChallengesActions,
  REGISTER_CHALLENGE_PARTICIPANT_FAIL
} from "store/actions/challenge/types";
import { toastr } from "react-redux-toastr";
import { withRouter, RouteComponentProps } from "react-router-dom";
import { ChallengeId, UserId } from "types";

interface FormData {}

interface OutterProps {
  userId: UserId;
}

type InnerProps = InjectedFormProps<FormData, Props> &
  RouteComponentProps<{ challengeId: ChallengeId }>;

type Props = InnerProps & OutterProps;

const CustomForm: FunctionComponent<Props> = ({ handleSubmit }) => (
  <Form onSubmit={handleSubmit} className="h-100">
    <Button.Submit
      color="primary"
      size="lg"
      className="text-uppercase h-100 w-100"
    >
      REGISTER
    </Button.Submit>
  </Form>
);

const enhance = compose<InnerProps, OutterProps>(
  withRouter,
  reduxForm<FormData, Props>({
    form: REGISTER_CHALLENGE_PARTICIPANT_FROM,
    onSubmit: async (_values, dispatch: any, { match }) =>
      await dispatch(
        registerChallengeParticipant(match.params.challengeId)
      ).then((action: ChallengesActions) => {
        switch (action.type) {
          case REGISTER_CHALLENGE_PARTICIPANT_FAIL: {
            return toastr.error(
              "ERROR",
              "An error occurred while registering as a challenge participant."
            );
          }
          default: {
            return Promise.resolve();
          }
        }
      })
  })
);

export default enhance(CustomForm);
