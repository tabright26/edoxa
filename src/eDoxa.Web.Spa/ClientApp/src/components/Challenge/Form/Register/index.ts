import ReduxForm from "./Register";

export default ReduxForm;
