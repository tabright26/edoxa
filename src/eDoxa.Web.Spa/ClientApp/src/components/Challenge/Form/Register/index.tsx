import React, { FunctionComponent } from "react";
import { Form } from "reactstrap";
import { reduxForm, InjectedFormProps } from "redux-form";
import Button from "components/Shared/Button";
import { REGISTER_CHALLENGE_PARTICIPANT_FROM } from "utils/form/constants";
import { compose } from "recompose";
import { registerChallengeParticipant } from "store/actions/challenge";
import { toastr } from "react-redux-toastr";
import { withRouter, RouteComponentProps } from "react-router-dom";
import { ChallengeId, UserId } from "types";

type FormData = {};

type InnerProps = InjectedFormProps<FormData, Props> &
  RouteComponentProps<{ challengeId: ChallengeId }>;

type OutterProps = {
  userId: UserId;
};

type Props = InnerProps & OutterProps;

const Register: FunctionComponent<Props> = ({ handleSubmit }) => (
  <Form onSubmit={handleSubmit} className="h-100">
    <Button.Submit
      color="primary"
      size="lg"
      className="text-uppercase h-100 w-100"
    >
      REGISTER
    </Button.Submit>
  </Form>
);

const enhance = compose<InnerProps, OutterProps>(
  withRouter,
  reduxForm<FormData, Props>({
    form: REGISTER_CHALLENGE_PARTICIPANT_FROM,
    onSubmit: (_values, dispatch, { match }) => {
      dispatch(registerChallengeParticipant(match.params.challengeId));
    },
    onSubmitFail: () => {
      toastr.error(
        "ERROR",
        "An error occurred while registering as a challenge participant."
      );
    }
  })
);

export default enhance(Register);
