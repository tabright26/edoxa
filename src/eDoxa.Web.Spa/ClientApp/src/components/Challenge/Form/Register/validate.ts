export const validate = values => {
  const errors: any = {};
  return errors;
};