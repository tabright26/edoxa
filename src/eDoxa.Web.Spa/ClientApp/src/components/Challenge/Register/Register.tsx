import React, { FunctionComponent, useState, useEffect } from "react";
import { Card } from "reactstrap";
import ChallengeForm from "components/Challenge/Form";
import authorizeService from "utils/oidc/AuthorizeService";
import { User } from "oidc-client";

interface Props {
  readonly className?: string;
  readonly canRegister: boolean;
}

const ChallengeRegister: FunctionComponent<Props> = ({
  className,
  canRegister
}) => {
  const [user, setUser] = useState<User>(null);
  useEffect(() => {
    authorizeService.getUser().then((user: User) => setUser(user));
  }, []);
  return (
    <>
      {canRegister && user && (
        <Card className={className}>
          <ChallengeForm.Register
            userId={user["sub"]}
            className="h-100 w-100"
          />
        </Card>
      )}
    </>
  );
};

export default ChallengeRegister;
