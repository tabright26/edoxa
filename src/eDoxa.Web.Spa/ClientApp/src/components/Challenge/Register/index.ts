import ChallengeRegister from "./Container";

export default ChallengeRegister;
