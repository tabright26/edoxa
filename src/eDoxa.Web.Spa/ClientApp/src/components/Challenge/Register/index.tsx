import React, { FunctionComponent } from "react";
import { Card, Button } from "reactstrap";
import ChallengeForm from "components/Challenge/Form";
import {
  HocUserProfileUserIdStateProps,
  withUserProfileUserId,
  withUserProfileGameIsAuthenticated,
  HocUserProfileGameIsAuthenticatedStateProps
} from "utils/oidc/containers";
import { compose } from "recompose";
import {
  ChallengeId,
  CHALLENGE_STATE_INSCRIPTION,
  Game,
  GameOptions
} from "types";
import { RouteComponentProps, withRouter } from "react-router-dom";
import { MapStateToProps, connect, DispatchProp } from "react-redux";
import { RootState } from "store/types";
import { show } from "redux-modal";
import { LINK_GAME_CREDENTIAL_MODAL } from "utils/modal/constants";

type Params = {
  readonly challengeId?: ChallengeId;
};

type OwnProps = RouteComponentProps<Params> &
  Params &
  HocUserProfileUserIdStateProps;

type StateProps = {
  readonly canRegister: boolean;
  readonly game: Game;
  readonly gameOptions: GameOptions;
};

type DispatchProps = {
  showLinkGameAccountCredentialModal: () => void;
};

type InnerProps = DispatchProp &
  DispatchProps &
  OwnProps &
  StateProps &
  HocUserProfileGameIsAuthenticatedStateProps;

type OutterProps = Params & {
  readonly className?: string;
};

type Props = InnerProps & OutterProps;

const Register: FunctionComponent<Props> = ({
  className,
  userId,
  isAuthenticated,
  canRegister,
  dispatch,
  gameOptions
}) =>
  canRegister &&
  (isAuthenticated ? (
    <ChallengeForm.Register className={className} userId={userId} />
  ) : (
    <Button
      color="primary"
      className={className}
      onClick={() =>
        dispatch(
          show(LINK_GAME_CREDENTIAL_MODAL, {
            gameOptions
          })
        )
      }
    >
      Link
    </Button>
  ));

const mapStateToProps: MapStateToProps<StateProps, OwnProps, RootState> = (
  state,
  ownProps
) => {
  const { data } = state.root.challenge;
  const challenge = data.find(
    challenge =>
      challenge.id ===
      (ownProps.match
        ? ownProps.match.params.challengeId
        : ownProps.challengeId)
  );
  return {
    canRegister:
      challenge.state === CHALLENGE_STATE_INSCRIPTION &&
      !challenge.participants.some(
        participant => participant.user.id === ownProps.userId
      ),
    game: challenge.game,
    gameOptions: state.static.games.games.find(x => x.name === challenge.game)
  };
};

const enhance = compose<InnerProps, OutterProps>(
  withRouter,
  withUserProfileUserId,
  connect(mapStateToProps),
  withUserProfileGameIsAuthenticated
);

export default enhance(Register);
