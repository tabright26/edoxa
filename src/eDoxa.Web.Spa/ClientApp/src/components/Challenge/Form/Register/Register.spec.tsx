import React from "react";
import Register from "./Register";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

it("renders correctly", () => {
  const store: any = {
    getState: () => {},
    dispatch: action => {},
    subscribe: () => {}
  };
  const tree = renderer
    .create(
      <Provider store={store}>
        <MemoryRouter>
          <Register userId="userId" />
        </MemoryRouter>
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
