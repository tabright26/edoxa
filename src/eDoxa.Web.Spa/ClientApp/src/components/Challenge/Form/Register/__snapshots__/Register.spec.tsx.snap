// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="h-100"
  onSubmit={[Function]}
>
  <button
    aria-label={null}
    className="text-uppercase undefined btn btn-primary btn-lg"
    onClick={[Function]}
    style={
      Object {
        "width": "75px",
      }
    }
    type="submit"
  >
    REGISTER
  </button>
</form>
`;
